src/csharp/Microsoft.Spark/ML/Feature/Pipeline.cs

```diff
@@ -70,10 +70,9 @@ public JavaPipelineStage[] GetStages()
             JavaPipelineStage[] result = new JavaPipelineStage[jvmObjects.Length];
             for (int i = 0; i < jvmObjects.Length; i++)
             {
-                (string constructorClass, string methodName) = DotnetUtils.GetUnderlyingType(jvmObjects[i]);
-                Type type = Type.GetType(constructorClass);
-                MethodInfo method = type.GetMethod(methodName, BindingFlags.NonPublic | BindingFlags.Static);
-                result[i] = (JavaPipelineStage)method.Invoke(null, new object[] { jvmObjects[i] });
+                var instance = DotnetUtils.ConstructInstanceFromJvmObject(
```

Comment:
instead of searching the assemblies `jvmObjects.Length` number of times, why not do the following instead ?
1) search the assemblies once
2) construct a dictionary that maps Java classes to Dotnet Classes
3) using constructed dictionary, iterate over the jvmObjects and construct the appropriate dotnet classes.